buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesTask.java

```diff
@@ -0,0 +1,125 @@
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Creates a plugin descriptor.
+ */
+public class PluginPropertiesTask extends Copy {
+
+    private PluginPropertiesExtension extension;
+
+    private File descriptorOutput = new File(getProject().getBuildDir(),
+            "generated-resources/plugin-descriptor.properties");
+
+    public PluginPropertiesTask() {
+        Project project = this.getProject();
+
+        File templateFile = new File(project.getBuildDir(), "templates/" + this.descriptorOutput.getName());
+
+        Task copyPluginPropertiesTemplate = project.getTasks().create("copyPluginPropertiesTemplate")
+                .doLast((unused) -> {
+                    InputStream resourceTemplate = PluginPropertiesTask.class.getResourceAsStream("/" + this.descriptorOutput.getName());
+                    templateFile.getParentFile().mkdirs();
+                    try {
+                        System.out.println(templateFile); ;
+                        Files.copy(resourceTemplate, templateFile.toPath(), StandardCopyOption.REPLACE_EXISTING);
+                    } catch (IOException e) {
+                        throw new RuntimeException(e);
+                    }
+                });
+        dependsOn(copyPluginPropertiesTemplate);
+        extension = project.getExtensions().create("esplugin", PluginPropertiesExtension.class, project);
+        project.afterEvaluate((unused) -> {
+            // check require properties are set
+            if (this.extension.getName() == null) {
+                throw new InvalidUserDataException("name is a required setting for esplugin");
+            }
+            if (this.extension.getDescription() == null) {
+                throw new InvalidUserDataException("description is a required setting for esplugin");
+            }
+            if (this.extension.getClassname() == null) {
+                throw new InvalidUserDataException("classname is a required setting for esplugin");
+            }
+
+            from(templateFile.getParentFile()).include(this.descriptorOutput.getName());
+            into(this.descriptorOutput.getParentFile());
+            Map<String, String> properties = this.generateSubstitutions();
+            expand(properties);
+            this.getInputs().properties(properties);
+        });
+
+    }
+
+    private Map<String, String> generateSubstitutions() {
+        Map<String, String> substitutions = new HashMap<>();
+        substitutions.put("name", this.extension.getName());
+        substitutions.put("description", this.extension.getDescription());
+        substitutions.put("version", this.extension.getVersion());
+        substitutions.put("elasticsearchVersion", Version.fromString(VersionProperties.getElasticsearch()).toString());
+
+        //TODO (Raf): No getter for project.getTargetCompatibility()? What's the alternative option for below line?
+        substitutions.put("javaVersion", String.valueOf(this.getProject().getVersion()));
```

Comment:
I could not find a getter for `project.getTargetCompatibility()`? Would appreciate feedback for `javaVersion` if `this.getProject().getVersion()` is not correct alternative.

Reply:
I think that's a java plugin conventions, you should be able to access it from Java in the same way we access source sets in existing code.